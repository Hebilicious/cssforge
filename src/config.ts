import type { ColorConfig } from "./modules/colors.ts";
import type { PrimitiveConfig } from "./modules/primitive.ts";
import type { SpacingConfig } from "./modules/spacing.ts";
import type { TypographyConfig } from "./modules/typography.ts";

/**
 * The main configuration object for CSSForge.
 */
export interface CSSForgeConfig {
  /**
   * Color configuration, including palettes, gradients, and themes.
   */
  colors: ColorConfig;
  /**
   * Typography configuration, including type scales and weights.
   */
  typography: TypographyConfig;
  /**
   * Spacing configuration for creating spacing scales.
   */
  spacing: SpacingConfig;
  /**
   * Primitive configuration for creating custom design tokens.
   */
  primitives: PrimitiveConfig;
}

/**
 * Define a CSSForge configuration.
 * @example
 * ```ts
 *  export default defineConfig(
    {
      spacing: {
        size: {
          value: {
            1: "0.25rem",
            2: "0.5rem",
            3: "0.75rem",
            4: "1rem",
          },
        },
      },
      typography: {
        fluid: {
          base: {
            value: {
              minWidth: 320,
              minFontSize: 14,
              minTypeScale: 1.25,
              maxWidth: 1435,
              maxFontSize: 16,
              maxTypeScale: 1.25,
              positiveSteps: 5,
              negativeSteps: 3,
            },
          },
        },
        weight: {
          arial: { value: { regular: "400", bold: "700" } },
        },
      },
      colors: {
        palette: {
          value: {
            coral: {
              100: { hex: "#FF7F50" },
            },
            mint: {
              100: { hex: "#4ADE80" },
            },
            indigo: {
              100: { hex: "#4F46E5" },
            },
          },
        },
      },
    },
  );
 * ```
 */
export function defineConfig<C extends CSSForgeConfig>(config: Partial<C>) {
  return config as C;
}
