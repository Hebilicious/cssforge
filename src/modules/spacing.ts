import type { UtopiaSpaceConfig } from "utopia-core";
import { calculateSpaceScale } from "utopia-core";
import { pxToRem, validateName } from "../helpers.ts";
import type { Output, ResolveMap } from "../lib.ts";

export interface FluidSpaceConfig {
  value: UtopiaSpaceConfig & {
    /**
     * An optional prefix for the generated CSS variables.
     */
    prefix?: string;
  };
}

type SpacingValue = string;

interface SpacingScale {
  [key: string]: SpacingValue;
}

/**
 * Settings for pixel to rem conversion.
 */
export interface PixelSettings {
  /**
   * Whether to convert pixel values to rems.
   * @default true
   */
  pxToRem?: boolean;
  /**
   * The base rem value to use for conversion.
   * @default 16
   */
  rem?: number;
}

/**
 * The spacing configuration object.
 */
export interface SpacingConfig {
  /**
   * Fluid spacing scales powered by utopia `calculateSpaceScale`.
   * Each key defines one scale. Tokens are derived from the returned sizes' labels.
   */
  fluid?: {
    [key: string]: FluidSpaceConfig;
  };
  /**
   * Static spacing scales defined under the `custom` key.
   * Each scale's `value` map contains token -> size (string) pairs.
   */
  custom?: {
    [key: string]: { value: SpacingScale; settings?: PixelSettings };
  };
}

/**
 * Convert spacing configuration into CSS custom properties.
 *
 * Generated variable naming:
 * - Custom scale: `--spacing-{scale}-{token}` (renamed from --space-)
 * - Fluid scale:  `--spacing_fluid-{prefixOrScale}-{token}` (renamed from --fluidspace-)

 *
 * @example
 * // Static + fluid combined
 * const { css } = processSpacing({
 *   fluid: {
 *     base: { value: { minSize: 4, maxSize: 24, minWidth: 320, maxWidth: 1280, negativeSteps: 0, positiveSteps: 2, prefix: "space" } }
 *   },
 *   custom: {
 *     gap: { value: { 1: "4px", 2: "8px" } }
 *   }
 * });
 */
export function processSpacing(spacing: SpacingConfig): Output {
  const cssOutput: string[] = [];
  const resolveMap: ResolveMap = new Map();

  if (spacing.fluid) {
    const moduleKey = "spacing_fluid";
    for (const [scaleName, { value }] of Object.entries(spacing.fluid)) {
      validateName(scaleName);
      const { prefix, ...utopiaConfig } = value;
      if (prefix) validateName(prefix);

      const { sizes, customPairs, oneUpPairs } = calculateSpaceScale(utopiaConfig);

      const resolvedPrefix = prefix ? `${scaleName}-${prefix}` : scaleName;
      for (const { label, clamp } of [sizes, customPairs, oneUpPairs].flat()) {
        const variableName = `--${moduleKey}-${resolvedPrefix}-${label}`;
        const cssVar = `${variableName}: ${clamp};`;
        cssOutput.push(cssVar);
        resolveMap.set(`${moduleKey}.${scaleName}@${label}`, {
          variable: cssVar,
          key: variableName,
          value: clamp,
        });
      }
    }
  }

  if (spacing.custom) {
    const moduleKey = "spacing";
    for (
      const [scaleName, { value, settings = { pxToRem: true, rem: 16 } }] of Object
        .entries(spacing.custom)
    ) {
      validateName(scaleName);
      for (const [scaleKey, scaleValue] of Object.entries(value)) {
        validateName(scaleKey);
        const convertedValue = settings.pxToRem
          ? pxToRem({ value: scaleValue, rem: settings.rem })
          : scaleValue;
        const varName = `--${moduleKey}-${scaleName}-${scaleKey}`;
        const variable = `${varName}: ${convertedValue};`;
        cssOutput.push(variable);
        resolveMap.set(`${moduleKey}.custom.${scaleName}.${scaleKey}`, {
          variable,
          key: varName,
          value: convertedValue,
        });
      }
    }
  }

  return { css: cssOutput.join("\n"), resolveMap };
}
