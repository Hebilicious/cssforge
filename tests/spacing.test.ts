// spacing_test.ts
import { assertEquals } from "jsr:@std/assert";
import { processSpacing } from "../src/modules/spacing.ts";
import { defineConfig } from "../src/config.ts";

Deno.test("processSpacing - generates correct spacing scale", () => {
  const config = defineConfig({
    spacing: {
      size: {
        value: { 1: "0.25rem", 2: "0.5rem", 3: "0.75rem", s: "16px" },
      },
    },
  });

  const result = processSpacing(config.spacing);
  const expected =
    "--size-1: 0.25rem;\n--size-2: 0.5rem;\n--size-3: 0.75rem;\n--size-s: 1rem;";
  assertEquals(result.css, expected);
});

// Test with string keys
Deno.test("processSpacing - handles settings", () => {
  const config = defineConfig({
    spacing: {
      size: {
        value: { 1: "16px", 2: "0.5rem" },
        settings: { pxToRem: true },
      },
      scale: {
        value: { "md": "8px", "lg": "0.75rem" },
        settings: { pxToRem: false },
      },
    },
  });

  const result = processSpacing(config.spacing);
  const expected =
    "--size-1: 1rem;\n--size-2: 0.5rem;\n--scale-md: 8px;\n--scale-lg: 0.75rem;";

  assertEquals(result.css, expected);
});
