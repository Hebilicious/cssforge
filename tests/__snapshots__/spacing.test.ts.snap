export const snapshot = {};

snapshot[`processSpacing - generates correct spacing scale 1`] = `
"--spacing-size-1: 0.25rem;
--spacing-size-2: 0.5rem;
--spacing-size-3: 0.75rem;
--spacing-size-s: 1rem;"
`;

snapshot[`processSpacing - generates correct spacing scale 2`] = `
[
  [
    "spacing.custom.size.1",
    {
      key: "--spacing-size-1",
      value: "0.25rem",
      variable: "--spacing-size-1: 0.25rem;",
    },
  ],
  [
    "spacing.custom.size.2",
    {
      key: "--spacing-size-2",
      value: "0.5rem",
      variable: "--spacing-size-2: 0.5rem;",
    },
  ],
  [
    "spacing.custom.size.3",
    {
      key: "--spacing-size-3",
      value: "0.75rem",
      variable: "--spacing-size-3: 0.75rem;",
    },
  ],
  [
    "spacing.custom.size.s",
    {
      key: "--spacing-size-s",
      value: "1rem",
      variable: "--spacing-size-s: 1rem;",
    },
  ],
]
`;

snapshot[`processSpacing - handles settings 1`] = `
"--spacing-size-1: 1rem;
--spacing-size-2: 0.5rem;
--spacing-scale-md: 8px;
--spacing-scale-lg: 0.75rem;"
`;

snapshot[`processSpacing - handles settings 2`] = `
[
  [
    "spacing.custom.size.1",
    {
      key: "--spacing-size-1",
      value: "1rem",
      variable: "--spacing-size-1: 1rem;",
    },
  ],
  [
    "spacing.custom.size.2",
    {
      key: "--spacing-size-2",
      value: "0.5rem",
      variable: "--spacing-size-2: 0.5rem;",
    },
  ],
  [
    "spacing.custom.scale.md",
    {
      key: "--spacing-scale-md",
      value: "8px",
      variable: "--spacing-scale-md: 8px;",
    },
  ],
  [
    "spacing.custom.scale.lg",
    {
      key: "--spacing-scale-lg",
      value: "0.75rem",
      variable: "--spacing-scale-lg: 0.75rem;",
    },
  ],
]
`;

snapshot[`processSpacing - generates fluid spacing (prefix) 1`] = `
"--spacing_fluid-base-foo-xs: clamp(0rem, 0rem + 0vw, 0rem);
--spacing_fluid-base-foo-s: clamp(0.25rem, -0.1667rem + 2.0833vw, 1.5rem);
--spacing_fluid-base-foo-m: clamp(0.25rem, -0.1667rem + 2.0833vw, 1.5rem);
--spacing_fluid-base-foo-l: clamp(0.5rem, -0.3333rem + 4.1667vw, 3rem);
--spacing_fluid-base-foo-xs-l: clamp(0rem, -1rem + 5vw, 3rem);
--spacing_fluid-base-foo-xs-s: clamp(0rem, -0.5rem + 2.5vw, 1.5rem);
--spacing_fluid-base-foo-s-m: clamp(0.25rem, -0.1667rem + 2.0833vw, 1.5rem);
--spacing_fluid-base-foo-m-l: clamp(0.25rem, -0.6667rem + 4.5833vw, 3rem);"
`;

snapshot[`processSpacing - generates fluid spacing (prefix) 2`] = `
[
  [
    "spacing_fluid.base@xs",
    {
      key: "--spacing_fluid-base-foo-xs",
      value: "clamp(0rem, 0rem + 0vw, 0rem)",
      variable: "--spacing_fluid-base-foo-xs: clamp(0rem, 0rem + 0vw, 0rem);",
    },
  ],
  [
    "spacing_fluid.base@s",
    {
      key: "--spacing_fluid-base-foo-s",
      value: "clamp(0.25rem, -0.1667rem + 2.0833vw, 1.5rem)",
      variable: "--spacing_fluid-base-foo-s: clamp(0.25rem, -0.1667rem + 2.0833vw, 1.5rem);",
    },
  ],
  [
    "spacing_fluid.base@m",
    {
      key: "--spacing_fluid-base-foo-m",
      value: "clamp(0.25rem, -0.1667rem + 2.0833vw, 1.5rem)",
      variable: "--spacing_fluid-base-foo-m: clamp(0.25rem, -0.1667rem + 2.0833vw, 1.5rem);",
    },
  ],
  [
    "spacing_fluid.base@l",
    {
      key: "--spacing_fluid-base-foo-l",
      value: "clamp(0.5rem, -0.3333rem + 4.1667vw, 3rem)",
      variable: "--spacing_fluid-base-foo-l: clamp(0.5rem, -0.3333rem + 4.1667vw, 3rem);",
    },
  ],
  [
    "spacing_fluid.base@xs-l",
    {
      key: "--spacing_fluid-base-foo-xs-l",
      value: "clamp(0rem, -1rem + 5vw, 3rem)",
      variable: "--spacing_fluid-base-foo-xs-l: clamp(0rem, -1rem + 5vw, 3rem);",
    },
  ],
  [
    "spacing_fluid.base@xs-s",
    {
      key: "--spacing_fluid-base-foo-xs-s",
      value: "clamp(0rem, -0.5rem + 2.5vw, 1.5rem)",
      variable: "--spacing_fluid-base-foo-xs-s: clamp(0rem, -0.5rem + 2.5vw, 1.5rem);",
    },
  ],
  [
    "spacing_fluid.base@s-m",
    {
      key: "--spacing_fluid-base-foo-s-m",
      value: "clamp(0.25rem, -0.1667rem + 2.0833vw, 1.5rem)",
      variable: "--spacing_fluid-base-foo-s-m: clamp(0.25rem, -0.1667rem + 2.0833vw, 1.5rem);",
    },
  ],
  [
    "spacing_fluid.base@m-l",
    {
      key: "--spacing_fluid-base-foo-m-l",
      value: "clamp(0.25rem, -0.6667rem + 4.5833vw, 3rem)",
      variable: "--spacing_fluid-base-foo-m-l: clamp(0.25rem, -0.6667rem + 4.5833vw, 3rem);",
    },
  ],
]
`;

snapshot[`processSpacing - fluid without prefix falls back to scale name 1`] = `
"--spacing_fluid-rhythm-xs: clamp(0rem, 0rem + 0vw, 0rem);
--spacing_fluid-rhythm-s: clamp(0.125rem, -0.25rem + 1.875vw, 1.25rem);
--spacing_fluid-rhythm-m: clamp(0.125rem, -0.25rem + 1.875vw, 1.25rem);
--spacing_fluid-rhythm-xs-s: clamp(0rem, -0.4167rem + 2.0833vw, 1.25rem);
--spacing_fluid-rhythm-s-m: clamp(0.125rem, -0.25rem + 1.875vw, 1.25rem);"
`;

snapshot[`processSpacing - fluid without prefix falls back to scale name 2`] = `
[
  [
    "spacing_fluid.rhythm@xs",
    {
      key: "--spacing_fluid-rhythm-xs",
      value: "clamp(0rem, 0rem + 0vw, 0rem)",
      variable: "--spacing_fluid-rhythm-xs: clamp(0rem, 0rem + 0vw, 0rem);",
    },
  ],
  [
    "spacing_fluid.rhythm@s",
    {
      key: "--spacing_fluid-rhythm-s",
      value: "clamp(0.125rem, -0.25rem + 1.875vw, 1.25rem)",
      variable: "--spacing_fluid-rhythm-s: clamp(0.125rem, -0.25rem + 1.875vw, 1.25rem);",
    },
  ],
  [
    "spacing_fluid.rhythm@m",
    {
      key: "--spacing_fluid-rhythm-m",
      value: "clamp(0.125rem, -0.25rem + 1.875vw, 1.25rem)",
      variable: "--spacing_fluid-rhythm-m: clamp(0.125rem, -0.25rem + 1.875vw, 1.25rem);",
    },
  ],
  [
    "spacing_fluid.rhythm@xs-s",
    {
      key: "--spacing_fluid-rhythm-xs-s",
      value: "clamp(0rem, -0.4167rem + 2.0833vw, 1.25rem)",
      variable: "--spacing_fluid-rhythm-xs-s: clamp(0rem, -0.4167rem + 2.0833vw, 1.25rem);",
    },
  ],
  [
    "spacing_fluid.rhythm@s-m",
    {
      key: "--spacing_fluid-rhythm-s-m",
      value: "clamp(0.125rem, -0.25rem + 1.875vw, 1.25rem)",
      variable: "--spacing_fluid-rhythm-s-m: clamp(0.125rem, -0.25rem + 1.875vw, 1.25rem);",
    },
  ],
]
`;

snapshot[`processSpacing - combines fluid and custom spacing 1`] = `
"--spacing_fluid-base-flux-xs: clamp(0rem, 0rem + 0vw, 0rem);
--spacing_fluid-base-flux-s: clamp(0.25rem, -0.1667rem + 2.0833vw, 1.5rem);
--spacing_fluid-base-flux-m: clamp(0.25rem, -0.1667rem + 2.0833vw, 1.5rem);
--spacing_fluid-base-flux-xs-s: clamp(0rem, -0.5rem + 2.5vw, 1.5rem);
--spacing_fluid-base-flux-s-m: clamp(0.25rem, -0.1667rem + 2.0833vw, 1.5rem);
--spacing-gap-1: 4px;
--spacing-gap-2: 8px;"
`;

snapshot[`processSpacing - combines fluid and custom spacing 2`] = `
[
  [
    "spacing_fluid.base@xs",
    {
      key: "--spacing_fluid-base-flux-xs",
      value: "clamp(0rem, 0rem + 0vw, 0rem)",
      variable: "--spacing_fluid-base-flux-xs: clamp(0rem, 0rem + 0vw, 0rem);",
    },
  ],
  [
    "spacing_fluid.base@s",
    {
      key: "--spacing_fluid-base-flux-s",
      value: "clamp(0.25rem, -0.1667rem + 2.0833vw, 1.5rem)",
      variable: "--spacing_fluid-base-flux-s: clamp(0.25rem, -0.1667rem + 2.0833vw, 1.5rem);",
    },
  ],
  [
    "spacing_fluid.base@m",
    {
      key: "--spacing_fluid-base-flux-m",
      value: "clamp(0.25rem, -0.1667rem + 2.0833vw, 1.5rem)",
      variable: "--spacing_fluid-base-flux-m: clamp(0.25rem, -0.1667rem + 2.0833vw, 1.5rem);",
    },
  ],
  [
    "spacing_fluid.base@xs-s",
    {
      key: "--spacing_fluid-base-flux-xs-s",
      value: "clamp(0rem, -0.5rem + 2.5vw, 1.5rem)",
      variable: "--spacing_fluid-base-flux-xs-s: clamp(0rem, -0.5rem + 2.5vw, 1.5rem);",
    },
  ],
  [
    "spacing_fluid.base@s-m",
    {
      key: "--spacing_fluid-base-flux-s-m",
      value: "clamp(0.25rem, -0.1667rem + 2.0833vw, 1.5rem)",
      variable: "--spacing_fluid-base-flux-s-m: clamp(0.25rem, -0.1667rem + 2.0833vw, 1.5rem);",
    },
  ],
  [
    "spacing.custom.gap.1",
    {
      key: "--spacing-gap-1",
      value: "4px",
      variable: "--spacing-gap-1: 4px;",
    },
  ],
  [
    "spacing.custom.gap.2",
    {
      key: "--spacing-gap-2",
      value: "8px",
      variable: "--spacing-gap-2: 8px;",
    },
  ],
]
`;
